// toggl-track/lib/timeEntry.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeEntry = void 0;
var TimeEntry = /** @class */ (function () {
    function TimeEntry(toggl) {
        this.toggl = toggl;
    }
    /**
     * GET TimeEntries​
     * Lists latest time entries.
     *
     * https://developers.track.toggl.com/docs/api/time_entry#get-timeentries
     */
    TimeEntry.prototype.list = function (query) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_c) {
                return [2 /*return*/, this.toggl.request('me/time_entries', {
                        query: {
                            before: query === null || query === void 0 ? void 0 : query.before,
                            since: query === null || query === void 0 ? void 0 : query.since,
                            start_date: (_a = query === null || query === void 0 ? void 0 : query.startDate) === null || _a === void 0 ? void 0 : _a.toString(),
                            end_date: (_b = query === null || query === void 0 ? void 0 : query.endDate) === null || _b === void 0 ? void 0 : _b.toString(),
                        },
                    })];
            });
        });
    };
    /**
     * GET Get current time entry​
     * Load running time entry for user ID.
     *
     * https://developers.track.toggl.com/docs/api/time_entry#get-get-current-time-entry
     */
    TimeEntry.prototype.current = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.toggl.request('me/time_entries/current')];
            });
        });
    };
    /**
     * POST TimeEntries​
     * Creates a new workspace TimeEntry.
     *
     * https://developers.track.toggl.com/docs/api/time_entry#post-timeentries
     */
    TimeEntry.prototype.create = function (workspaceId, body) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.toggl.request("workspaces/".concat(workspaceId, "/time_entries"), {
                        method: 'POST',
                        body: body,
                    })];
            });
        });
    };
    /**
     * PATCH Bulk editing time entries​
     * In short: http://tools.ietf.org/html/rfc6902 and http://tools.ietf.org/html/rfc6901 with some additions. Patch will be executed partially when there are errors with some records. No transaction, no rollback.
     *
     * https://developers.track.toggl.com/docs/api/time_entry#patch-bulk-editing-time-entries
     */
    TimeEntry.prototype.updateBulk = function (timeEntryIds, workspaceId, body) {
        return __awaiter(this, void 0, void 0, function () {
            var timeEntryIdsString;
            return __generator(this, function (_a) {
                timeEntryIdsString = timeEntryIds.join(',');
                return [2 /*return*/, this.toggl.request("workspaces/".concat(workspaceId, "/time_entries/").concat(timeEntryIdsString), {
                        method: 'PATCH',
                        body: body,
                    })];
            });
        });
    };
    /**
     * PUT TimeEntries​
     * Updates a workspace time entry.
     *
     * https://developers.track.toggl.com/docs/api/time_entry#put-timeentries
     */
    TimeEntry.prototype.update = function (timeEntryId, workspaceId, body) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.toggl.request("workspaces/".concat(workspaceId, "/time_entries/").concat(timeEntryId), {
                        method: 'PUT',
                        body: body,
                    })];
            });
        });
    };
    /**
     * DELETE TimeEntries​
     * Deletes a workspace time entry.
     *
     * https://developers.track.toggl.com/docs/api/time_entry#delete-timeentries
     */
    TimeEntry.prototype.delete = function (timeEntryId, workspaceId) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.toggl.request("workspaces/".concat(workspaceId, "/time_entries/").concat(timeEntryId), {
                        method: 'DELETE',
                    })];
            });
        });
    };
    return TimeEntry;
}());
exports.TimeEntry = TimeEntry;
